### PR TITLE
feat:  add graphql introspection middleware

### PR DESCRIPTION
**Pull request by spetrunin**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jensneuse** [comment]
Don't forget to update the docs

File: packages/sdk/src/definition/federation.ts [hunk 1]
```diff
@@ -13,15 +14,26 @@ export const isFederationService = (schema: GraphQLSchema): boolean => {
...
+
+	let opts: AxiosRequestConfig = {
 		headers: {
 			'Content-Type': 'application/json',
 			Accept: 'application/json',
 		},
-	});
+	};
+
+	if (requestMiddleware) {
```
Comment (StarpTech) [inline packages/sdk/src/definition/federation.ts:32]: I read the issue, but I'm still not convinced why we need to provide a middleware mechanism. Could you elaborate?

File: packages/sdk/src/definition/index.ts [hunk 2]
```diff
@@ -248,6 +248,12 @@ export interface DataSource<Custom = unknown> {
 	Directives: DirectiveConfiguration[];
 }
 
+export type RequestMiddleware = (requestConfig: AxiosRequestConfig) => AxiosRequestConfig;
```
Comment (StarpTech) [inline packages/sdk/src/definition/index.ts:251]: Axios is an implementation detail. In the future, we might switch the client. Don't expose the type to the user.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/sdk/src/definition/index.ts (hunk 2 above)
Comment (spetrunin) [reply packages/sdk/src/definition/index.ts:251]: I need a detailed specification in this case - like what should be possible

File: packages/sdk/src/definition/federation.ts (hunk 1 above)
Comment (spetrunin) [reply packages/sdk/src/definition/federation.ts:32]: At the moment there is only one way to authenticate an introspection request - add a StaticHeader 
but this means that such header will be used also for all requests

So we have a need in some way to provide various auth options for the introspection calls
open to any proposals

File: packages/sdk/src/definition/index.ts (hunk 2 above)
Comment (jensneuse) [reply packages/sdk/src/definition/index.ts:251]: I actually think that at this moment, a function that takes the headers and allows us to override them should be enough.

Comment (spetrunin) [reply packages/sdk/src/definition/index.ts:251]: could be there a basic auth?

Comment (jensneuse) [reply packages/sdk/src/definition/index.ts:251]: yes, it's possible, but currently we don't have this use case
maybe you find a way to keep it flexible enough but keeping it simple?

File: packages/sdk/src/definition/federation.ts (hunk 1 above)
Comment (StarpTech) [reply packages/sdk/src/definition/federation.ts:32]: What about?

```ts
const spacex = introspect.graphql({
	apiNamespace: 'spacex',
	url: 'https://api.spacex.land/graphql/',
	headers: (builder) => builder.addClientRequestHeader('X-Authorization', 'Authorization'),
	introspection: {
		headers: (builder) => builder.addClientRequestHeader('X-Authorization', 'Authorization'),
	}
});
```

My favorite API would look like this :smile: Exposing a builder or middleware is powerful but I see no reason to add this complexity here. It's also not simple to use.

```ts
const spacex = introspect.graphql({
	apiNamespace: 'spacex',
	url: 'https://api.spacex.land/graphql/',
	headers: [
		Headers.static('X-Authorization', 'Authorization'),
		Headers.client('X-Authorization', 'Authorization')],
	introspection: {
		headers: [
			Headers.static('X-Authorization', 'Authorization'),
			Headers.auth('username', 'password')],
	},
});
```

**Review by jensneuse — APPROVED** [review]
LGTM. I've only added a few minor changes in the docs.

File: docs-website/config/tags.js [hunk 3]
```diff
@@ -281,7 +281,7 @@ const tags = {
 	'configure mTLS':
 		'/docs/wundergraph-config-ts-reference/configure-mtls-for-http-based-data-sources',
 	'configure introspection for protected API':
-		'/docs/wundergraph-config-ts-reference/configure-introspection#introspection-of-protected-graphql-api',
+		'/docs/wundergraph-config-ts-reference/configure-introspection#introspection-of-protected-graph-ql-ap-is',
```
Comment (StarpTech) [inline docs-website/config/tags.js:284]: `graph-ql`

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: docs-website/config/tags.js (hunk 3 above)
Comment (spetrunin) [reply docs-website/config/tags.js:284]: it is how nextjs generated an anchor from `## heading`

Comment (spetrunin) [reply docs-website/config/tags.js:284]: https://github.com/wundergraph/wundergraph/pull/214/files#diff-a88d471cf72fc2ee1fffbe5d4cf67db38d8c45ffc3e5e702cfd9281a598361eeR81

header is here

**Review by StarpTech — APPROVED** [review]
Great work!

**Pull request merged**